cmd/bisync/bisync_test.go

```diff
@@ -1897,3 +1899,13 @@ func checkError(t *testing.T, err error, msgAndArgs ...interface{}) {
 	}
 	assert.NoError(t, err, msgAndArgs...)
 }
+
+// for example, replaces TestS3{juk_h}:dir with TestS3,directory_markers=true:dir
+// because NewFs needs the latter
+func (b *bisyncTest) replaceHex(remote string) string {
+	if bilib.HasHexString(remote) {
+		remote = strings.ReplaceAll(remote, fs.ConfigString(b.parent1), fs.ConfigStringFull(b.parent1))
```

Comment:
Just note that `ConfigStringFull` often contains credentials so it does **not** want to go in any logs.

So I think this is probably not a good idea...

What are you trying to fix?

Reply:
Hmm, ok. The problem I'm trying to solve is this, let me know if there's a better solution:

The test case scenarios have lots of globs like
`copy-as {datadir/}file1.txt {path1/}subdir file20.txt`
which get translated into instructions to set the stage before the test.

The `{path1/}` variable usually ends up looking like `TestS3:rclone-test-qemaser8watigiw0haguhij9/bisync-013220-biqaq/normalization/path1/`, which is fine, as we can just pass that directly into `cache.Get` or `fs.NewFs` and all will be well.

But if the config is overridden then it will have a hexstring, like `TestS3{juk_h}:rclone-test-qemaser8watigiw0haguhij9/bisync-013220-biqaq/normalization/path1/`. In that case, we can't just pass it in directly, because `NewFs` will complain about `config name contains invalid characters`, because of those braces. Previously we just stripped the hexstring off, but this was really not a good idea, as then the Fs returned did not have the right options (for example, `directory_markers` changes whether or not we can do the `createemptysrcdirs` test.)

`NewFs` will happily accept `TestS3,directory_markers=true:dir`, so I am looking for a way to replace `TestS3{juk_h}:dir` with `TestS3,directory_markers=true:dir`. Or a different way of reverse engineering a `TestS3{juk_h}:dir` string into an `fs` with the right config.

By the way, is it intentional that `cache.Get` can store a canonical name if given an FsString but can't accept a canonical name to look up a cached Fs? That was surprising to me.

Reply:
> Just note that `ConfigStringFull` often contains credentials so it does **not** want to go in any logs.

OK, I think I was being overly alarmist there!

Yes it is true that lots of rclone users construct on the fly remotes with secrets in, and these **will** appear in the output of `ConfigStringFull`, so we should try not to log those.

**However** none of the test remotes do that so I think it is perfectly safe to do this in test code only and a quick skim convinces me that your change is only used in test code. And in fact we do use this already in the b2 and s3 internal tests for an almost identical reason.

So with that proviso, I think what you've got is OK, and many apologies for overstating the problem!

> By the way, is it intentional that cache.Get can store a canonical name if given an FsString but can't accept a canonical name to look up a cached Fs? That was surprising to me.

I made the cache mostly for the rc/api so backends don't get created and destroyed too often. So its purpose was to cache things by the name the user gave it, and generally users don't change the name they use.

We could make the cache look up things by canonical name and that might not be a bad idea, but I suspect it would be of limited use outside the tests.

Reply:
Ok, thanks for clarifying! You are right that this is only used on the tests, so I think it's ok. But good to know about `ConfigStringFull` for my future knowledge.

> We could make the cache look up things by canonical name and that might not be a bad idea, but I suspect it would be of limited use outside the tests.

There are times when it might be useful for `chroot`-ing an Fs, but I think your proposal on #1644 is a more efficient solution to that problem.